idx/memory/memory.go

```diff
@@ -1543,18 +1541,32 @@ func (m *UnpartitionedMemoryIdx) Delete(orgId uint32, pattern string) ([]idx.Arc
 			}
 		}
 	}()
+
+	m.RLockHigh()
 	tree, ok := m.tree[orgId]
 	if !ok {
+		m.RUnlockHigh()
 		return nil, nil
 	}
 	found, err := find(tree, pattern)
+	m.RUnlockHigh()
 	if err != nil {
 		return nil, err
 	}
 
-	for _, f := range found {
-		deleted := m.delete(orgId, f, true, true)
+	bc := BlockContext{}
+
+	// limit delete operations holding a write lock to 200ms of every second
+	tl := NewTimeLimiter(time.Second, time.Millisecond*200, time.Now())
+
+	for _, node := range found {
+		tl.Wait()
+		lockStart := time.Now()
+		bc = m.Lock()
```

Comment:
Maybe I'm misunderstanding something here, but it seems like the write lock is being acquired and released in each iteration. Perhaps it should only be released when it has used up it's time slice?

Repeatedly acquiring write locks is very expensive in the memory index.

Reply:
I am still experimenting with deleting by leaf. I don't think this current approach solves the entire problem set. If someone tries to delete 'toplevel.*' when they have a lot of child nodes it will still lock up the index since the current method would recursively call all the way down the tree.

Reply:
That is correct, deleting `*` would be even worse.

We definitely need the delete call to find all leaf nodes to be deleted while only holding read locks.  Then delete these in batches with write locks.